Mastering Directus: A Comprehensive Guide to Aggregation and Grouping Filters (#29)
directus-community/guest-authoring: post/index.md

---
title: "Mastering Directus: A Comprehensive Guide to Aggregation and Grouping Filters"
description: "120-160 characters"
author:
    name: "jay bharadia"
    avatar_file_name: "my-avatar.jpg"
---

> ## Introduction
> 
> In the ever-evolving landscape of web development, the need for robust content management systems (CMS) has become more apparent than ever.
> 
> -   Directus stands out as a powerful choice, offering a myriad of features to handle data intricacies.
> -   However, navigating the complexities of advanced filtering, especially with nesting and logical operators, can be a daunting task.
> 
> Join us as we embark on a hands-on exploration, providing practical examples, insightful strategies, and a roadmap to elevate your Directus proficiency.
> 
> Welcome to "Mastering Directus Advanced Filtering" – your comprehensive guide to navigating the depths of Directus with confidence and efficiency.
> 
> ## Before You Start
> 
> > ℹ️ This guide uses the new [Directus SDK 11](https://docs.directus.io/guides/sdk/getting-started.html). Please refer docs if you are using an older version.
> 
> ### Consider we have Blog Website
> 
> -   In our blog website, we have articles list with filters
> 
> ### Article List
> 
> ![Article List!](./images/all-articles-list.png)
> 
> ### Filters
> 
> ![Article List!](./images/filters.png)
> 
> > ## Real World Requirement
> 
> -   We need to show articles as per the **filters** are selected ( we have month , year filter ) and **search** and **sort**
> -   Get data Month wise , Year wise , also `when month and year both are selected`
> -   Get Data when search value is entered
> -   when combination of above is happened ( month + year + search )
> -   show Count of data as per the filters

Reviewer: ```suggestion
Directus is a really powerful data platform that automatically generates REST and GraphQL APIs, and an SDK to make working with them more convenient. 

There's a number of powerful query parameters to filter, sort, limit, and apply functions to the data returned. But with any powerful tool, complex usage can require a learning curve. 

In this post, we'll cover some common and powerful advanced querying that you can use in your applications. In this example, imagine a blog with a search and some time-based filters. Note that each filter has a number of posts that match the query.

![A list of blog posts with a sidebar showing a search, a year filter with 2022 and 2023 checkboxes, and the same for months. Each year and month has a number in brackets next to it.](https://github.com/directus-community/guest-authoring/assets/1461554/e6e229bb-13e1-4a54-bfe7-623c12425bb3)

Before diving into queries, set up a new collection called `articles` with `title` and `published_date` fields. Any extra fields are fine, but we won't use them today. Then create 10 example items that we can use to test our queries. 

In your Public role permissions, allow read access to the `articles` collection.
```

---

### Directus Schema Setup

-   Set up your schema as shown in screenshot

![Blog Schema](./images/blog-schema.png)

-   Now create `published_date` with `date` format

![Article List!](./images/published-date-field.png)

### Accessing Items ( JS Part)

> > Items are individual pieces of data in your database. They can be anything, from articles, to IoT status checks. Learn more about [Items](https://docs.directus.io/user-guide/overview/glossary.html#items).

Reviewer: We can remove this because it's summarized in the short paragraph in the last suggestion.

> ### Initialize Directus instance
> 
> ```js
> import { createDirectus, rest, aggregate, readItems } from "@directus/sdk";
> 
> const client = createDirectus("https://directus.example.com").with(rest());
> ```

Reviewer: ````suggestion
### Initialize the Directus SDK

Create a new directory and open it in both your code editor and a terminal. In your terminal, run `npm init -y` to create a `package.json` file. Then, run `npm install @directus/sdk` to install the Directus SDK.

Create an `index.js` file and initialize the SDK, replacing the example URL with your Directus Project URL:

```js
import { createDirectus, rest, aggregate, readItems } from "@directus/sdk";

const client = createDirectus("https://directus.example.com").with(rest());
```
````

> ### Get Data by Month
> 
> ```js
>   restClient
>         .request(
>             readItems("articles", {
>                 filter: {
>                     _or: [
>                         {
>                             "month(published_date)": {
>                                 // January
>                                 _eq: 1,
>                             },
>                         },
>                         {
>                             "month(published_date)": {
>                                 // Feb
>                                 _eq: 2,
>                             },
>                         },
>                         {
>                             "month(published_date)": {
>                                 // March
>                                 _eq: 3,
>                             },
>                         },
>                     ],
>                 },
>             })
>         )
> 
> // Response
> [{...}] // total length 1
> ```

Reviewer: ````suggestion
### Get Data By Month

The following filter will return any item in the `articles` collection that is published in months 1, 2, or 3 (January, February, or March). 

This filter uses the `_or` [filter operator](https://docs.directus.io/reference/filter-rules.html#filter-operators) which contains an array. If at least one of the object conditions are `true`, it will be included.

`month()` is a [DateTime function](https://docs.directus.io/reference/query.html#datetime-functions) that will extract the month from a datetime, date, or timestamp field and return the month number. 

```js
client.request(
      readItems("articles", {
          filter: {
              _or: [
                  {
                      "month(published_date)": {
                          // January
                          _eq: 1,
                      },
                  },
                  {
                      "month(published_date)": {
                          // Feb
                          _eq: 2,
                      },
                  },
                  {
                      "month(published_date)": {
                          // March
                          _eq: 3,
                      },
                  },
              ],
          },
      })
  );
```

Save the file, and run `node index.js` to try it out and see an array of matching items. Of course, you require at least one item which matches the criteria to receive a non-empty array.
````

(I'll fix code indentation when publishing)

> ### Get Data By Year
> 
> ```js
>   restClient
>         .request(
>             readItems("articles", {
>                 filter: {
>                     _or: [
>                             {
>                                 "year(published_date)": {
>                                         _eq: 2023,
>                                 },
>                             },
>                             {
>                                 "year(published_date)": {
>                                         _eq: 2022,
>                                 },
>                             },
>                     ],
>                 },
>             })
>         )
> // Response
> [{...},{...}] // Total count 2
> ```

Reviewer: ````suggestion
### Get Data By Year

The same approach can be used with any other DateTime functions, such as `year()`:

```js
  client.request(
            readItems("articles", {
                filter: {
                    _or: [
                            {
                                "year(published_date)": {
                                        _eq: 2023,
                                },
                            },
                            {
                                "year(published_date)": {
                                        _eq: 2022,
                                },
                            },
                    ],
                },
            })
        );
```

:::info Dynamic Variables

Directus also has a set of [dynamic variables](https://docs.directus.io/reference/filter-rules.html#dynamic-variables) that can be used in filtering, including `$NOW` and `$NOW(<adjustment>)` (for example `$NOW(-1 week)`), that can be used for relative filters.

:::
````

### Get Data By Month & Year & search

> 
> ```js
>    restClient
>         .request(
>             readItems("articles", {
>                 filter: {
>                     // Notice below _and -> logical operator
>                     _and: [
>                         {
>                             // Month
>                             _or: [
>                                 {
>                                     "month(published_date)": {
>                                         // January
>                                         _eq: 1,
>                                     },
>                                 },
>                                 {
>                                     "month(published_date)": {
>                                         // Feb
>                                         _eq: 2,
>                                     },
>                                 },
>                                 {
>                                     "month(published_date)": {
>                                         // March
>                                         _eq: 3,
>                                     },
>                                 },
>                             ],
>                         },
>                         {
>                             // Year
>                             _or: [
>                                 {
>                                     "year(published_date)": {
>                                         _eq: 2023,
>                                     },
>                                 },
>                                 {
>                                     "year(published_date)": {
>                                         _eq: 2022,
>                                     },
>                                 },
>                             ],
>                         },
>                         {
>                             // Title search
>                             title: {
>                                 _contains: "gpt",
>                             },
>                         },
>                     ],
>                 },
>             })
>         )
> 
> //Response
> [{...},{...}] // total count 2
> ```

Reviewer: ````suggestion
### Combining Filters: Date & Search Term

Filter syntax can get confusing as you build increasingly complex requests. However, with what we've learnt so far, we can create something very powerful. 

This filter starts with an `_and` logical operator - this means that every member of the array must be true for a match, and there are three:

1. Must be published in January, February, or March.
2. Must be published in 2022 or 2023.
3. The title must contain 'Hello'.

Notice that for both of the month-based filters, it begins with an `_or` logical operator. Only one of these must match for the whole item to count for the `_and` operator. 

```js
   client.request(
            readItems("articles", {
                filter: {
                    _and: [
                        {
                            _or: [
                                {
                                    "month(published_date)": {
                                        _eq: 1,
                                    },
                                },
                                {
                                    "month(published_date)": {
                                        _eq: 2,
                                    },
                                },
                                {
                                    "month(published_date)": {
                                        _eq: 3,
                                    },
                                },
                            ],
                        },
                        {
                            _or: [
                                {
                                    "year(published_date)": {
                                        _eq: 2022,
                                    },
                                },
                                {
                                    "year(published_date)": {
                                        _eq: 2023,
                                    },
                                },
                            ],
                        },
                        {
                            title: {
                                _contains: "Hello",
                            },
                        },
                    ],
                },
            })
        )
```
````

## Aggregation & Grouping

> 

Reviewer: I would like to see you write an introduction to Aggregation & Grouping here, not a straight copy from the docs. Perhaps in this context you can find an interesting way to explain it.

> ### Get Filter Count By Year
> 
> ```js
> // Get articles count by year for showing in filters
> const result = await client.request(
> 	aggregate('articles', {
> 		aggregate: { count: '*' },
>         groupBy: ['year(published_date)']
> 	})
> );
> 
> // Response
> {
>   "data": [
>     {
>       "published_date_year": 2022,
>       "count": 35
>     },
>     {
>       "published_date_year": 2023,
>       "count": 44
>     }
>   ]
> }
> ```
> 
> ![Filter Year Count](./images/filter-year-count.jpeg)

Reviewer: ````suggestion
### Count Articles By Year

In this example, we will combine aggregation and grouping. Aggregation allows us to perform a count on all articles that exist. The result will then be displayed for each bucket as grouped by the year value.

```js
const result = await client.request(
	aggregate('articles', {
		aggregate: { count: '*' },
    groupBy: ['year(published_date)']
	})
);

// Response
{
  "data": [
    {
      "published_date_year": 2022,
      "count": 35
    },
    {
      "published_date_year": 2023,
      "count": 44
    }
  ]
}
```

You can use the `month()` DateTime function to perform the same logic for the month filters. 
````

### Get Filter Count By Month

```js
// Get articles count by month for showing in filters
const result = await client.request(
	aggregate('articles', {
		aggregate: { count: '*' },
        groupBy: ['month(published_date)']
	})
);

// Response
{
  "data": [
    {
      "published_date_month": 1,
      "count": 4
    },
    {
      "published_date_month": 2,
      "count": 16
    },
    {
      "published_date_month": 3,
      "count": 15
    },
    {
      "published_date_month": 4,
      "count": 9
    },
    {
      "published_date_month": 5,
      "count": 8
    },
    {
      "published_date_month": 6,
      "count": 4
    },
    {
      "published_date_month": 7,
      "count": 2
    },
    {
      "published_date_month": 8,
      "count": 4
    },
    {
      "published_date_month": 9,
      "count": 2
    },
    {
      "published_date_month": 10,
      "count": 5
    },
    {
      "published_date_month": 11,
      "count": 6
    },
    {
      "published_date_month": 12,
      "count": 6
    }
  ]
}
```

![Month Count](./images/month-filter-count.jpeg)

-   Below image shows few more options

>     ![Article List!](./images/all-filters-options.png)

Reviewer: Let's remove this. It's repetitive. We've already demonstrated month and year are interchangable logic-wise.

## Summary

> -   By delving into this comprehensive guide, you have not only master advanced filtering in Directus but also learned valuable insights into SDK 11 integration and real-world application development.
> -   This guide serves as a cornerstone for creating powerful queries, providing a solid foundation for addressing complex use cases with confidence.

Reviewer: ```suggestion
The examples given above use hard-coded values, but with little additional work in your real-world application, you can create dynamic queries that returns just the data your users want.

By following this guide based on a real-world example, you have not only learnt how to combine filters, but also been introduced to logical operators, DateTime functions, and explored aggregation & grouping.
```
